modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/other/ReadOnlySecondaryUserStoreAPITokenGenerationTestCase.java

```diff
@@ -0,0 +1,213 @@
+
+        //create a new gateway session cookie since server is restarted
+        String newGatewaySessionCookie = createSession(gatewayContext);
+
+        //check for AxisFault when adding roles to the read only user store
+        testUserMgtClient = new UserManagementClient(backendURL, newGatewaySessionCookie);
+        AuthenticatorClient testAuthenticatorClient = new AuthenticatorClient(gatewayContext.getContextUrls().getBackEndUrl());
+
+        try{
+            testUserMgtClient.addRole("CAT.COM/testRole", null, new String[]{"/permission/admin/login"
+                    , "/permission/admin/manage/api/subscribe"});
+
+        }catch (AxisFault axisFault) {
+            log.info("The Exception is expected since adding roles is disabled when user store is readonly.");
+        }
+        Assert.assertEquals(testUserMgtClient.roleNameExists("CAT.COM/testRole"), false,
+                "User Store is still not readonly");
+
+        //Load the back-end dummy API
+        loadSynapseConfigurationFromClasspath("artifacts" + File.separator + "AM"
+                + File.separator + "synapseconfigs" + File.separator + "rest"
+                + File.separator + "dummy_api.xml", gatewayContext, newGatewaySessionCookie);
+
+        //add an API
+        String apiContext = "test";
+        String tags = "testing";
+        String url = gatewayUrls.getWebAppURLNhttp() + "response";
+        String description = "This is a test API created by API manager integration test";
+
+        apiPublisher = new APIPublisherRestClient(publisherUrls.getWebAppURLHttp());
+        apiStore = new APIStoreRestClient(storeUrls.getWebAppURLHttp());
+        apiProvider = publisherContext.getSuperTenant().getContextUser().getUserName();
+
+        apiPublisher.login("admin", "admin");
+
+        APIRequest apiRequest = new APIRequest(API_NAME, apiContext, new URL(url));
+        apiRequest.setTags(tags);
+        apiRequest.setDescription(description);
+        apiRequest.setVersion(API_VERSION);
+        apiRequest.setProvider(apiProvider);
+        apiRequest.setSandbox(url);
+
+        apiPublisher.addAPI(apiRequest);
+
+        //publish the API
+        APILifeCycleStateRequest updateRequest = new APILifeCycleStateRequest(API_NAME, apiProvider,
+                APILifeCycleState.PUBLISHED);
+        apiPublisher.changeAPILifeCycleStatus(updateRequest);
+
+        authenticatorClient.logOut();
+
+        //go to store with external store user credentials
+        apiStore.login(newUserName, newUserPassword);
+
+        //create application
+        apiStore.addApplication(APP_NAME, "Unlimited", "some_url", "NewApp");
+
+        //subscribed the API
+        SubscriptionRequest subscriptionRequest = new SubscriptionRequest(API_NAME, apiProvider);
+        subscriptionRequest.setApplicationName(APP_NAME);
+
+        apiStore.subscribe(subscriptionRequest);
+
+        //Generate sandbox Token
```

Comment:
This sort of comments generally indicates the need of splitting the contents of the logic into smaller pieces and having them declared as separate methods for better readability. For example, you can easily grab this token generation portion out and put into a separate method.